### DISCUSSION TITLE
Convergence Test Question in CTHYB

### DISCUSSION THREAD

**Discussion by andrewkhardy** [TRIQS/triqs#895] [General]
Hello, I would like to perform some simple convergence tests on the ModelDMFT tutorials but I am struggling to access the Green's function object for CTHYB. 

From [here](https://triqs.github.io/cthyb/latest/reference/solver.html) if I call
`>> Solver.G_iw['up']`
I get a response like 

> Greens Function G_up with mesh Matsubara Freq Mesh of size 2050, Domain: Matsubara domain with beta = 10, statistic = Fermion, positive_only : 0 and target_shape (1, 1): 
> , dtype=object

Is there any way to extract from this object a numpy array of the actual values? 

I would like to compute the error, such as the difference in G from one iteration to another as a function of things like the number of iterations, number of sweeps in the QMC, number of Matsubara frequencies, etc.  Is there an easier way other than converting to a numpy array?

**Comment by the-hampel** [comment]
Hi @andrewkhardy,

yes there is. `.data` on the Gf object (not the BlockGf) will give you a view on the stored data a numpy array: https://triqs.github.io/triqs/unstable/documentation/manual/triqs/gfs/py/block/GfImFreq.html

**Reply by andrewkhardy** [reply]
@the-hampel Thank you so much for the quick response to a naive question. That works thanks. 

For my second question, it looks like there's a lot stored in the CTHYB solver, but I did not see anything in `S.performance_analysis`. Does it store error per sweep or anything like that, or what should I use to determine the number of sweeps? Perhaps I'm coming at this wrong as I'm used to thinking of classical MC where it's easy to see the energy plateau as it converges.

**Reply by the-hampel** [reply]
that is a bit more tricky in this case. triqs/cthyb does not provide directly error metrics. A good place to start is this tutorial: https://triqs.github.io/cthyb/unstable/guide/cthyb_convergence_tests.html , which will guide you through the different convergence metrics you could look at. 

As soon as you are looking into a full DMFT calculation you can also look at more involved metrics like the DMFT self-consistency condition. But for the solution of the AIM model itself the tutorial above should summarize the most important metrics.

**Reply by andrewkhardy** [reply]
As always, I should have read the documentation more carefully, but thank you for being patient.

**Reply by andrewkhardy** [reply]
@the-hampel I've started running my own checks here to compare to CT-INT and my first confusion is that running the linked tutorial gives me an autocorrelation time of zero. This also happens with Model-DMFT tutorials and my own scripts. I know it's not a problem with storing data because the CT-HYB terminal output will give something like 

> Average sign: 0.9988

>Average order: 0.8517

>Auto-correlation time: 0

Is it just not being computed correctly, or have I botched something? I can only get nonzero if I run an absurdly small number of cycles

**Reply by the-hampel** [reply]
Are you using the current 3.1.x release of triqs? If so the explanation is quite simple. In the 3.1.x release of triqs the auto correlation time measurement was still rather simplistic and only considered how the `sign` of the MC chain de-correlates. However, if your sign is by construction 1, e.g. in the model dmft tutorial for a simple 1 band bethe lattice, there is not much de-correlation to measure if the variable is always 1. Hence, cthyb will give as autocorrelation time 0. 

We fixed this in the current unstable branch of cthyb by also using the perturbation order additionally for the auto correlation time estimate. This will be part of the upcoming triqs 3.2. release. 

So for now, please don't be worried if the value is 0 in system that report a sign of 1 (please check if that is the case). 

Best,
Alex
